spark/src/main/scala/org/apache/spark/sql/delta/DeltaLog.scala

```diff
@@ -185,17 +185,33 @@ class DeltaLog private(
    * ------------------ */
 
   /**
-   * Returns a new [[OptimisticTransaction]] that can be used to read the current state of the
-   * log and then commit updates. The reads and updates will be checked for logical conflicts
-   * with any concurrent writes to the log.
+   * Returns a new [[OptimisticTransaction]] that can be used to read the current state of the log
+   * and then commit updates. The reads and updates will be checked for logical conflicts with any
+   * concurrent writes to the log, and post-commit hooks can be used to notify the table's catalog
+   * of schema changes, etc.
    *
    * Note that all reads in a transaction must go through the returned transaction object, and not
    * directly to the [[DeltaLog]] otherwise they will not be checked for conflicts.
+   *
+   * @param catalogTableOpt The [[CatalogTable]] for the table this transaction updates. Passing
+   * None asserts this is a path-based table with no catalog entry.
+   *
+   * @param snapshotOpt THe [[Snapshot]] this transaction should use, if not latest.
    */
-  def startTransaction(): OptimisticTransaction = startTransaction(None)
+  def startTransaction(
+      catalogTableOpt: Option[CatalogTable],
```

Comment:
I know this is already merged, but DeltaLog has a forTable overload that accepts CatalogTable, couldnt we store it in a DeltaLog field?

Reply:
Great question!

`DeltaLog` is shared between all threads/sessions on the cluster, and is also a cached object that can go away at any time. Further, one physical Delta table can be registered multiple times with multiple catalogs, e.g. if two users create different external table entries for the same path. So if we rely on the `DeltaLog` to carry the correct catalog entry, we risk finding somebody else's catalog entry (because they updated it more recently than we did), or even not finding one at all (because of a cache eviction since we last looked). The latter is especially a risk for some of the older V1-style commands, that repeatedly lookup the `DeltaLog` based on path, instead of remembering the original catalog lookup. Because of all that, I went with `OptimisticTransaction` as safer, because it's session-private. From there, we just have to plumb through the catalog table that was found during analysis -- usually from a `DeltaTableV2` or `LogicalRelation`.